File: src/data/dataset.py
from data.pattern import Pattern
import numpy as np


class Dataset:
    __size = 0
    __num_dim = 0  # number of attributes
    __num_classes = 0
    __patterns = None

    def __init__(self, size, n_dim, c_num, patterns):
        if size <= 0 or n_dim <= 0 or c_num <= 0:
            raise ValueError("Incorrect input data set information")
        self.__size = size
        self.__num_dim = n_dim
        self.__num_classes = c_num
        self.__patterns = np.array(patterns)

    def get_pattern(self, index):
        return self.__patterns[index]

    def get_patterns(self):
        return self.__patterns

    def __str__(self):
        if len(self.__patterns) == 0:
            return "null"
        txt = f"{self.__size}, {self.__num_dim}, {self.__num_classes}\n"
        for pattern in self.__patterns:
            txt += f"{pattern}\n"
        return txt

    def get_num_dim(self):
        return self.__num_dim

    def get_num_classes(self):
        return self.__num_classes

    def get_size(self):
        return self.__size

File: src/mofgbmlpy/core/data/dataset.cpp
#include "dataset.hpp"
#include <stdexcept>

Dataset::Dataset(int size, int num_dim, int num_classes, const std::vector<Pattern*>& patterns)
    : size(size), num_dim(num_dim), num_classes(num_classes), patterns(patterns) {
    if (size <= 0 || num_dim <= 0 || num_classes <= 0) {
        throw std::invalid_argument("size, num_dim and num_classes must be positive");
    }

    if (size != static_cast<int>(patterns.size())) {
        throw std::invalid_argument("Size doesn't match patterns vector size");
    }
    Pattern first_pattern = *patterns[0];
    if (num_dim != first_pattern.get_num_dim()) {
        throw std::invalid_argument("Number of dimensions does not match the first pattern's dimensions");
    }
}

Dataset::Dataset(const Dataset& other)
    : size(other.size), num_dim(other.num_dim), num_classes(other.num_classes), patterns(other.patterns) {
}

Pattern* Dataset::get_pattern(int index) const {
    return patterns.at(index);
}

const std::vector<Pattern*>& Dataset::get_patterns() const {
    return patterns;
}

int Dataset::get_num_dim() const {
    return num_dim;
}

int Dataset::get_num_classes() const {
    return num_classes;
}

int Dataset::get_size() const {
    return size;
}

Dataset::operator std::string() const
{
    if (patterns.empty()) {
        return "null";
    }
    std::string txt = std::to_string(size) + ", " + std::to_string(num_dim) + ", " + std::to_string(num_classes) + "\n";
    for (const auto& pattern : patterns) {
        txt += std::string(*pattern) + "\n";
    }
    return txt;
}

bool Dataset::operator==(const Dataset& other) const
{
    if (size == other.size &&
        num_dim == other.num_dim &&
        num_classes == other.num_classes) {
        for (int i = 0; i < get_size(); ++i) {
            if (!(*(patterns[i]) == *(other.patterns[i]))) {
                return false;
            }
        }
        return true;
    }
    return false;
}

Dataset* Dataset::clone() const
{
    return new Dataset(*this);
}

std::string Dataset::to_string() const {
    return static_cast<std::string>(*this);
}
